src/transformStateWithClones.js

```
@@ -7,7 +7,40 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let prevState = { ...state };
+
+  for (const action of actions) {
+    let stateCopy = { ...prevState };
+
+    switch (action.type) {
+      case 'addProperties':
+        if (
+          action.extraData &&
+          typeof action.extraData === 'object' &&
+          !Array.isArray(action.extraData)
+        ) {
+          stateCopy = { ...stateCopy, ...action.extraData };
+        }
+        break;
+      case 'removeProperties':
+        if (Array.isArray(action.keysToRemove)) {
+          for (const key of action.keysToRemove) {
+            delete stateCopy[key];
+          }
+        }
+        break;
+      case 'clear':
+        stateCopy = {};
+        break;
+      default:
+        break;
```

This violates checklist item #3: 'switch/case should always have default case for error handling'. While a `default` case is present, it's empty. This means any unknown action types will be silently ignored. It's better to throw an error here to alert developers of potential typos or incorrect actions.